iphone/Classes/KrollContext.m

```diff
@@ -1529,3 +1548,80 @@ - (void)dealloc
 }
 
 @end
+
+#ifdef USE_JSCORE_FRAMEWORK
+
+@implementation JSTimerManager
+
+- (instancetype)initInContext:(JSContext *)context
+{
+  self = [super init];
+  if (!self) {
+    return nil;
+  }
+
+  self.timers = [NSMutableDictionary new];
+
+  int (^setInterval)(void) = ^() {
+    return [self setIntervalFromArguments:JSContext.currentArguments shouldRepeat:YES];
+  };
+  context[@"setInterval"] = setInterval;
+
+  int (^setTimeout)(void) = ^() {
+    return [self setIntervalFromArguments:JSContext.currentArguments shouldRepeat:NO];
+  };
+  context[@"setTimeout"] = setTimeout;
+
+  void (^clearInterval)(JSValue *) = ^(JSValue *value) {
+    return [self clearIntervalWithIdentifier:value.toInt32];
+  };
+  context[@"clearInterval"] = clearInterval;
+  context[@"clearTimeout"] = clearInterval;
+
+  return self;
+}
+
+- (void)dealloc
+{
+  RELEASE_TO_NIL(self.timers);
+
+  [super dealloc];
+}
+
+- (void)invalidateAllTimers
+{
+  for (NSNumber *timerIdentifier in self.timers.allKeys) {
+    [self clearIntervalWithIdentifier:timerIdentifier.intValue];
+  }
+}
+
+- (int)setIntervalFromArguments:(NSArray<JSValue *> *)arguments shouldRepeat:(BOOL)shouldRepeat
+{
+  NSMutableArray *callbackArgs = [arguments.mutableCopy autorelease];
+  JSValue *callbackFunction = [callbackArgs objectAtIndex:0];
+  [callbackArgs removeObjectAtIndex:0];
+  double interval = [[callbackArgs objectAtIndex:0] toDouble] / 1000;
+  [callbackArgs removeObjectAtIndex:0];
+  NSNumber *timerIdentifier = [NSNumber numberWithUnsignedInteger:[[[[NSUUID alloc] init] autorelease] hash]];
+  NSTimer *timer = [NSTimer scheduledTimerWithTimeInterval:interval
+                                                   repeats:shouldRepeat
+                                                     block:^(NSTimer *_Nonnull timer) {
+                                                       [callbackFunction callWithArguments:callbackArgs];
```

Comment:
I tested with Kroll thread and timers are not properly firing here. I think it's because we are permanently blocking the run loop with our infinite while loop. For now i only made the new timer management available when using jscore *and* running on main thread.